runtime/resetting_histogram.go

```diff
@@ -0,0 +1,298 @@
+}
+
+// Snapshot resets the timer and returns a read-only copy of its sorted contents.
+func (t *StandardResettingHistogram) Snapshot() Histogram {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	currentValues := t.values
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
+
+	sort.Slice(currentValues, func(i, j int) bool { return currentValues[i] < currentValues[j] })
+
+	return &ResettingHistogramSnapshot{
+		values: currentValues,
+	}
+}
+
+func (t *StandardResettingHistogram) Clear() {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.values = make([]int64, 0, InitialResettingHistogramSliceCap)
+}
+
+// Record the duration of an event.
+func (t *StandardResettingHistogram) Update(d int64) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.values = append(t.values, d)
+}
+
+// ResettingHistogramSnapshot is a point-in-time copy of another ResettingHistogram.
+type ResettingHistogramSnapshot struct {
+	values              []int64
+	mean                float64
+	thresholdBoundaries []float64
+	calculated          bool
+}
+
+// Snapshot returns the snapshot.
+func (t *ResettingHistogramSnapshot) Snapshot() Histogram { return t }
+
+func (*ResettingHistogramSnapshot) Update(int64) {
+	panic("Update called on a ResettingHistogramSnapshot")
+}
+
+func (t *ResettingHistogramSnapshot) Clear() {
+	panic("Clear called on a ResettingHistogramSnapshot")
+}
+
+func (t *ResettingHistogramSnapshot) Sample() Sample {
+	panic("Sample called on a ResettingHistogramSnapshot")
+}
+
+func (t *ResettingHistogramSnapshot) Count() int64 { return int64(len(t.values)) }
+
+// Values returns all values from snapshot.
+func (t *ResettingHistogramSnapshot) Values() []int64 {
+	return t.values
+}
+
+func (t *ResettingHistogramSnapshot) Min() int64 {
+	if len(t.values) > 0 {
+		return t.values[0]
+	}
+	return 0
+}
+
+func (t *ResettingHistogramSnapshot) Variance() float64 {
+	if 0 == len(t.values) {
+		return 0.0
+	}
+	m := t.Mean()
+	var sum float64
+	for _, v := range t.values {
+		d := float64(v) - m
+		sum += d * d
+	}
+	return sum / float64(len(t.values))
+}
+
+func (t *ResettingHistogramSnapshot) Max() int64 {
+	if len(t.values) > 0 {
+		return t.values[len(t.values)-1]
+	}
+	return 0
+}
+
+func (t *ResettingHistogramSnapshot) StdDev() float64 {
+	return math.Sqrt(t.Variance())
+}
+
+func (t *ResettingHistogramSnapshot) Sum() int64 {
+	var sum int64
+	for _, v := range t.values {
+		sum += v
+	}
+	return sum
+}
+
+// Percentile returns the boundaries for the input percentiles.
+func (t *ResettingHistogramSnapshot) Percentile(percentile float64) float64 {
+	t.calc([]float64{percentile})
+
+	return t.thresholdBoundaries[0]
+}
+
+// Percentiles returns the boundaries for the input percentiles.
+func (t *ResettingHistogramSnapshot) Percentiles(percentiles []float64) []float64 {
+	t.calc(percentiles)
+
+	return t.thresholdBoundaries
+}
+
+// Mean returns the mean of the snapshotted values
+func (t *ResettingHistogramSnapshot) Mean() float64 {
+	if !t.calculated {
```

Comment:
This is racy and rather weird, because the mean and the state is over-written on every re-calculation, and the input for the recalculation can change. I suggest you only store the mean in the state, because it's the only value that gets re-used from multiple getters. 

The threshold boundaries should not be part of the state because they don't get reused.

Reply:
Fixed. Generally you have one actor interacting with a snapshot, but I agree that better to make this non-racy.